### DISCUSSION TITLE
Better integration with modern .NET practices

### DISCUSSION THREAD

**Discussion by ah1508** [nhibernate/nhibernate-core#3266] [Ideas]
EF Core is catching up and even if Nhibernate still have unique and more advanced features, the lack of 

* built-in integration with modern .NET
* mapping by code documentation (already mentionned in another issue). Until now the most reliable and comprehensive documentation of mapping by code http://notherdev.blogspot.com/2012/02/nhibernates-mapping-by-code-summary.html

does not help onboarding for new comers.

Suggestions:

The `ConventionalModelMapper` could register `ClassMapping` by introspection, like for hbm.xml files.

It would avoid this kind of boilerplate:

```csharp
public class MyMapper : ConventionModelMapper
{
    public MyMapper()
    {
        GetType().Assembly.GetTypes().Where(t =>
            t.BaseType != null &&
            t.BaseType.IsGenericType &&
            t.BaseType.GetGenericTypeDefinition() == typeof(ClassMapping<>)
        ).ToList().ForEach(AddMapping);
}
```

The integration with modern dotnet standards (ILogger, IConfiguration) could be built-in. The default setup still relies on hibernate.cfg.xml while `IConfiguration` abstraction has beeen the standard for many years now. 

Coupling with `ILogger` requires an extension (`NHibernate.Extensions.Logging`) while it could also be built-in.

To sum up, I'd love to see that:

```csharp
ISessionFactory sessionFactory = ISessionFactory.Create(configuration, loggerFactory, "myassembly");
```

where `configuration` is an instance of `IConfiguration` with a `NHibernate` section.

Implémentation proposal of this method:

```csharp
    public static ISessionFactory Create(IConfiguration configuration, ILoggerFactory loggerFactory, string assembly, Action<NHibernate.Cfg.Configuration>? customizer = null)
    {
        return Create(configuration, loggerFactory, new string[] { assembly }, customizer);
    }

    public static ISessionFactory Create(IConfiguration configuration, ILoggerFactory loggerFactory, string[] assemblies, Action<NHibernate.Cfg.Configuration>? customizer = null)
    {
        var logger = NHibernateLogger.For(typeof(ISessionFactory));
        logger.Info("Configure logging: NHibernate will use Microsoft.Extensions.Logging");
        NHibernateLogger.SetLoggersFactory(new MicrosoftLoggerFactory(loggerFactory)); // equivalent : loggerFactory.UseAsNHibernateLoggerProvider();
        
        logger.Info("Get properties from NHibernate configuration section");
        IDictionary<string, string> properties = configuration.GetSection("NHibernate")
           .GetChildren()
           .Where(x => x.Value != null)
           .ToDictionary(x => x.Key, x => x.Value!);

        if (properties.TryGetValue(NHibernate.Cfg.Environment.ConnectionStringName, out var connStringName) && !string.IsNullOrWhiteSpace(connStringName))
        {
            logger.Info("Property {0} references connection string {1}", NHibernate.Cfg.Environment.ConnectionStringName, connStringName);
            var connectionString = configuration.GetConnectionString(connStringName) ?? throw new Exception(connStringName + " does not exists");
            logger.Info("Add property {0}", NHibernate.Cfg.Environment.ConnectionString);
            properties[NHibernate.Cfg.Environment.ConnectionString] = connectionString;
            properties.Remove(NHibernate.Cfg.Environment.ConnectionStringName);
        }
        var cfg = new NHibernate.Cfg.Configuration { Properties = properties };

        foreach (var assembly in assemblies.Select(Assembly.Load))
        {
            logger.Info("Register {0}", assembly.GetName());
            cfg.AddAssembly(assembly); // for hbm.xml files

            assembly.GetTypes()
                .Where(t => typeof(ModelMapper).IsAssignableFrom(t))
                .Select(t => Activator.CreateInstance(t))
                .OfType<ModelMapper>()
                .ToList()
                .ForEach(mapper =>
                {
                    logger.Info("Register {0} as {0}", mapper.GetType().Name, nameof(ModelMapper));
                    cfg.AddMapping(mapper.CompileMappingForAllExplicitlyAddedEntities());
                });
        }
        if (customizer != null)
        {
            logger.Info("Apply {0} on configuration", nameof(customizer));
            customizer(cfg);
        }
        logger.Info("Create SessionFactory");
        var sessionFactory = cfg.BuildSessionFactory();
        logger.Info("SessionFactory ready");
        return sessionFactory;
    }
```

I also think that a "transactional session context" should be added (see https://github.com/nhibernate/nhibernate-core/issues/1907#issuecomment-827428731). `TransactionScope` is a bit forgotten but remains a neutral way to surround a unit of work. It does not dispose asynchronously but all other implementations of `ICurrentSessionContext` have also their drawbacks.

My point: it's a pity if NHibernate, which is not as popular as EFCore despite several more advanced features, gives the feeling to be behind modern dotnet because of a few details.

**Comment by hazzik** [comment]
Thanks for your input @ah1508, however, that is more a discussion than an issue as this is not really actionable.

**Comment by fredericDelaporte** [comment]
Theses features should be considered for 6.0 I think.

**Comment by knuxbbs** [comment]
I agree with all the points bring by the OP. Especially the lack of documentation for mapping by code.

How could we help with this, specifically?

**Reply by fredericDelaporte** [reply]
Documentation can be contributed in the core repository in docbook format. It is in the [`doc` sub-folder](https://github.com/nhibernate/nhibernate-core/tree/master/doc). The end result will be published on next minor as "The NHibernate Reference" [here](https://nhibernate.info/doc/index.html).

**Comment by ah1508** [comment]
How about contacting the author of http://notherdev.blogspot.com/2012/02/nhibernates-mapping-by-code-summary.html ? Maybe some parts must be rephrased, but it's better than writing everything from scratch. For the time being it is the most accurate source of documentation.
